File: agentic-clj/src/agentic/workflow_timers.clj
(ns agentic.workflow-timers
  "Workflow `timers` (spec/v1/primitives.md, section 8) and the two clocks they read. Processing time
   is a logical clock owned by the system: `system-clock` reads wall time, `manual-clock` moves only
   when told to (`advance!`) and never backwards, which is how a conformance fixture's
   `advance_time_ms` drives it. Event time is the conversation watermark, the highest `event_time_ms`
   turn metadata seen, folded from the log.

   Pending timers are never held in memory across turns: `fold` derives them from `timer_scheduled`
   minus `timer_fired`, so a restarted runtime recovers exactly the timers still owed and can neither
   re-schedule nor double-fire one. Every turn_received of a workflow with timers records the
   processing clock reading (`processing_time_ms`), so the clock itself is recoverable from the log."
  (:require [agentic.cep-fold :as cep-fold]
            [agentic.context :as ctx]))

;; ---- clocks ----

(defprotocol Clock
  (now-ms [_] "The current processing time in milliseconds."))

(defrecord SystemClock []
  Clock
  (now-ms [_] (System/currentTimeMillis)))

(defrecord ManualClock [!now]
  Clock
  (now-ms [_] @!now))

(defn system-clock [] (->SystemClock))

(defn manual-clock
  "A logical processing clock starting at `start-ms` (default 0)."
  ([] (manual-clock 0))
  ([start-ms]
   (when (neg? start-ms)
     (throw (ex-info (str "logical time cannot be negative: " start-ms) {:error/class :fatal})))
   (->ManualClock (atom (long start-ms)))))

(defn advance!
  "Move a manual clock forward by `ms`; logical time never moves backwards."
  [^ManualClock clock ms]
  (when (neg? ms)
    (throw (ex-info (str "logical time never moves backwards: advance by " ms) {:error/class :fatal})))
  (swap! (:!now clock) + (long ms))
  clock)

;; ---- the fold ----

(def empty-state
  {:watermark-ms nil :processing-time-ms nil :pending {} :fired []})

(defn fold
  "Timer state as a fold over a conversation's events: the watermark, the last recorded processing
   clock reading, the pending timers by id and the fired ids in firing order."
  [events]
  (reduce (fn [state {:keys [type payload]}]
            (case type
              :turn-received
              (cond-> state
                (some? (:event-time-ms payload))
                (update :watermark-ms #(max (or % (:event-time-ms payload)) (:event-time-ms payload)))
                (some? (:processing-time-ms payload))
                (assoc :processing-time-ms (:processing-time-ms payload)))
              :timer-scheduled
              (assoc-in state [:pending (:timer-id payload)]
                        {:timer-id (:timer-id payload) :clock (:clock payload) :due-ms (:due-ms payload)})
              :timer-fired
              (if (contains? (:pending state) (:timer-id payload))
                (-> state
                    (update :pending dissoc (:timer-id payload))
                    (update :fired conj (:timer-id payload)))
                state)
              state))
          empty-state
          events))

(defn recovered-processing-time
  "The processing clock reading a restarted runtime rebuilds from its logs: the highest recorded
   `processing_time_ms`, or 0 when no turn recorded one."
  [event-logs]
  (reduce max 0 (keep #(:processing-time-ms (fold %)) event-logs)))

;; ---- turns ----

(defn event-time-of
  "The turn's `event_time_ms` metadata as a long, or nil when the turn carries none; the same read
   the CEP fold uses (`agentic.cep-fold/event-time-ms`), so timers and patterns agree on event time."
  [event]
  (cep-fold/event-time-ms (:metadata event)))

(defn watermark-after
  "The watermark once a turn with `event-time` (or nil) has arrived: never lower than before."
  [state event-time]
  (cond (nil? event-time) (:watermark-ms state)
        (nil? (:watermark-ms state)) event-time
        :else (max (:watermark-ms state) event-time)))

(defn- reading
  "The clock a timer reads: processing time from the system clock, event time from the watermark
   (0 until the conversation has seen an event time)."
  [clock processing-now watermark]
  (if (= "event" clock) (or watermark 0) processing-now))

(defn- active?
  "Timers only take part in a turn when the workflow declares some and the context carries a clock
   and the conversation's prior events (`:clock`, `:prior-events`, set by agentic.core)."
  [graph c]
  (and (seq (:timers graph)) (some? (:clock c))))

(defn processing-now
  "The processing clock reading for this turn, or nil when the workflow declares no timers."
  [graph c]
  (when (active? graph c) (now-ms (:clock c))))

(defn received-payload
  "The `turn_received` payload with the event time and, for a workflow with timers, the processing
   clock reading this turn was processed at."
  [graph event c payload]
  (let [event-time (event-time-of event)
        now (processing-now graph c)]
    (cond-> payload
      (some? event-time) (assoc :event-time-ms event-time)
      (some? now) (assoc :processing-time-ms now))))

(defn fire-due!
  "Before a later turn is received: fire every pending timer whose clock reads at or past its
   deadline, ordered by (due_ms, timer_id), appending `timer_fired` and calling the timer's tool
   with its declared payload. The first turn of a conversation fires nothing."
  [graph event c]
  (when (active? graph c)
    (let [state (fold (:prior-events c))
          first-turn? (empty? (:prior-events c))
          now (now-ms (:clock c))
          watermark (watermark-after state (event-time-of event))]
      (when-not first-turn?
        (doseq [{:keys [timer-id due-ms]}
                (->> (vals (:pending state))
                     (filter (fn [t] (>= (reading (:clock t) now watermark) (:due-ms t))))
                     (sort-by (juxt :due-ms :timer-id)))]
          (ctx/emit! c :timer-fired {:timer-id timer-id :due-ms due-ms})
          (let [spec (some #(when (= timer-id (:id %)) %) (:timers graph))]
            (when (:tool spec)
              (ctx/call-tool c (:tool spec) (or (:payload spec) {})))))))))

(defn schedule!
  "After the first turn of a conversation is received: append `timer_scheduled` for every declared
   timer, due `after_ms` past its clock's current reading."
  [graph event c]
  (when (and (active? graph c) (empty? (:prior-events c)))
    (let [state (fold (:prior-events c))
          now (now-ms (:clock c))
          watermark (watermark-after state (event-time-of event))]
      (doseq [{:keys [id after-ms clock]} (:timers graph)]
        (let [clock (or clock "processing")]
          (ctx/emit! c :timer-scheduled {:timer-id id :clock clock
                                         :due-ms (+ (reading clock now watermark) (long after-ms))}))))))

File: ports/pyagentic/agentic/workflow_timers.py
"""Workflow `timers` (spec/v1/primitives.md, section 8) and the two clocks they read.

Processing time is a logical clock the runtime owns: `wall_clock_ms` reads the system clock,
`ManualClock` moves only when told to (`advance`) and never backwards, which is how a
conformance fixture's `advance_time_ms` drives it. Event time is the conversation watermark,
the highest `event_time_ms` turn metadata seen, folded from the log.

Pending timers are never held in memory across turns: `TimerState.fold` derives them from
`timer_scheduled` minus `timer_fired`, so a restarted runtime recovers exactly the timers
still owed and can neither re-schedule nor double-fire one. Every `turn_received` of a
workflow with timers records the processing clock reading (`processing_time_ms`), so the
clock itself is recoverable from the log (`ManualClock.recovered_from`).
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, Sequence

from .cep import event_time_ms
from .errors import ValidationError
from .events import Event, EventLog, Turn

PROCESSING = "processing"
EVENT = "event"


class ManualClock:
    """Logical processing time in milliseconds; callable like any `Clock`."""

    def __init__(self, start_ms: int = 0) -> None:
        if start_ms < 0:
            raise ValidationError(f"logical time cannot be negative: {start_ms}")
        self._now = int(start_ms)
        self._guard = threading.Lock()

    def __call__(self) -> int:
        with self._guard:
            return self._now

    def advance(self, ms: int) -> int:
        if ms < 0:
            raise ValidationError(f"logical time never moves backwards: advance by {ms}")
        with self._guard:
            self._now += int(ms)
            return self._now

    @classmethod
    def recovered_from(cls, log: EventLog) -> "ManualClock":
        """The clock a restarted runtime starts from: the highest processing time any
        conversation's `turn_received` recorded, or zero when none did."""
        recorded = 0
        for conversation_id in log.conversation_ids():
            state = TimerState.fold(log.read(conversation_id))
            if state.processing_time_ms is not None:
                recorded = max(recorded, state.processing_time_ms)
        return cls(recorded)


def event_time_of(turn: Turn) -> Optional[int]:
    """`metadata.event_time_ms` as an int, None when the turn has none: the same single read the CEP
    fold uses (:func:`agentic.cep.event_time_ms`), so timers and patterns agree on event time."""
    return event_time_ms(turn.metadata)


@dataclass(frozen=True)
class PendingTimer:
    timer_id: str
    clock: str
    due_ms: int


@dataclass
class TimerState:
    """What the log says about a conversation's clocks and timers."""

    watermark_ms: Optional[int] = None
    processing_time_ms: Optional[int] = None
    pending: Dict[str, PendingTimer] = field(default_factory=dict)
    fired: List[str] = field(default_factory=list)

    @classmethod
    def fold(cls, events: Iterable[Event]) -> "TimerState":
        state = cls()
        for event in events:
            payload = event.payload
            if event.type == "turn_received":
                if payload.get("event_time_ms") is not None:
                    t = int(payload["event_time_ms"])
                    state.watermark_ms = t if state.watermark_ms is None else max(state.watermark_ms, t)
                if payload.get("processing_time_ms") is not None:
                    state.processing_time_ms = int(payload["processing_time_ms"])
            elif event.type == "timer_scheduled":
                timer_id = str(payload["timer_id"])
                state.pending[timer_id] = PendingTimer(timer_id, str(payload.get("clock", PROCESSING)),
                                                      int(payload["due_ms"]))
            elif event.type == "timer_fired":
                timer_id = str(payload["timer_id"])
                if state.pending.pop(timer_id, None) is not None:
                    state.fired.append(timer_id)
        return state

    def watermark_after(self, event_time: Optional[int]) -> Optional[int]:
        """The watermark once a turn carrying `event_time` has arrived: never lower than before."""
        if event_time is None:
            return self.watermark_ms
        if self.watermark_ms is None:
            return event_time
        return max(self.watermark_ms, event_time)


def _reading(clock: str, processing_now: int, watermark: Optional[int]) -> int:
    if clock == EVENT:
        return watermark if watermark is not None else 0
    if clock == PROCESSING:
        return processing_now
    raise ValidationError(f"unknown timer clock {clock!r}; one of processing, event")


Invoke = Callable[[str, Mapping[str, Any]], Any]
Append = Callable[[str, Mapping[str, Any]], Any]


class WorkflowTimers:
    """The workflow's `timers` block, evaluated at the boundary of every turn.

    `fire_due` runs before a later turn's `turn_received` and fires every pending timer whose
    clock has reached its deadline, ordered by `(due_ms, timer_id)`, invoking the timer's tool
    with its declared payload. `schedule` runs after the first turn's `turn_received` and
    schedules every declared timer against the clock it names. Both are no-ops for a workflow
    without timers, which then also records no clock readings.
    """

    def __init__(self, specs: Sequence[Mapping[str, Any]], clock: Callable[[], int]) -> None:
        self.specs: List[Mapping[str, Any]] = list(specs or [])
        self.clock = clock
        for spec in self.specs:
            _reading(str(spec.get("clock", PROCESSING)), 0, 0)  # rejects an unknown clock at build time

    @property
    def active(self) -> bool:
        return bool(self.specs)

    def received_payload(self, turn: Turn, payload: Mapping[str, Any]) -> Dict[str, Any]:
        """`turn_received` payload with `event_time_ms` (when the turn carries one) and, for a
        workflow with timers, the processing clock reading."""
        out = dict(payload)
        event_time = event_time_of(turn)
        if event_time is not None:
            out["event_time_ms"] = event_time
        if self.active:
            out["processing_time_ms"] = self.clock()
        return out

    def fire_due(self, turn: Turn, prior: Sequence[Event], append: Append, invoke: Invoke) -> List[str]:
        """Fire the timers due at the head of this turn; returns their ids in firing order."""
        if not self.active or not prior:
            return []
        state = TimerState.fold(prior)
        now = self.clock()
        watermark = state.watermark_after(event_time_of(turn))
        due = sorted((t for t in state.pending.values() if _reading(t.clock, now, watermark) >= t.due_ms),
                     key=lambda t: (t.due_ms, t.timer_id))
        fired: List[str] = []
        for pending in due:
            append("timer_fired", {"timer_id": pending.timer_id, "due_ms": pending.due_ms})
            fired.append(pending.timer_id)
            spec = next((s for s in self.specs if s["id"] == pending.timer_id), None)
            if spec is not None and spec.get("tool"):
                invoke(str(spec["tool"]), dict(spec.get("payload") or {}))
        return fired

    def schedule(self, turn: Turn, prior: Sequence[Event], append: Append) -> List[PendingTimer]:
        """Schedule every declared timer on the conversation's first turn."""
        if not self.active or prior:
            return []
        state = TimerState.fold(prior)
        now = self.clock()
        watermark = state.watermark_after(event_time_of(turn))
        scheduled: List[PendingTimer] = []
        for spec in self.specs:
            clock = str(spec.get("clock", PROCESSING))
            timer = PendingTimer(str(spec["id"]), clock, _reading(clock, now, watermark) + int(spec["after_ms"]))
            append("timer_scheduled", {"timer_id": timer.timer_id, "clock": timer.clock, "due_ms": timer.due_ms})
            scheduled.append(timer)
        return scheduled
